docs/container.md

```diff
@@ -0,0 +1,313 @@
+
+  ```php
+  use Piwik\Translation\Translator;
+  use Piwik\Plugins\MyPlugin\Dao\MyEntityDao;
+
+  class API
+  {
+      /**
+       * @var Translator
+       */
+      private $translator;
+
+      /**
+       * @var MyEntityDao
+       */
+      private $myEntityDao;
+
+      public function __construct(Translator $translator, MyEntityDao $myEntityDao)
+      {
+          $this->translator = $translator;
+          $this->myEntityDao = $myEntityDao;
+      }
+
+      // ...
+  }
+  ```
+
+3. **Determine if your class has any non-class dependencies.**
+
+  Sometimes your class may depend on data that is in the DI container and not just classes, for example, [path.tmp](https://github.com/piwik/piwik/blob/master/config/global.php).
+
+  If your class has such a dependency, add it to the constructor without a type and then manually configure the parameter in DI config. _See the next section to learn how to create DI config._
+
+  For example:
+
+  ```php
+  use Piwik\Translation\Translator;
+  use Piwik\Plugins\MyPlugin\Dao\MyEntityDao;
+
+  class API
+  {
+      /**
+       * @var Translator
+       */
+      private $translator;
+
+      /**
+       * @var MyEntityDao
+       */
+      private $myEntityDao;
+
+      /**
+       * @var string
+       */
+      private $tmpPath;
+
+      public function __construct(Translator $translator, MyEntityDao $myEntityDao, $tmpPath)
+      {
+          $this->translator = $translator;
+          $this->myEntityDao = $myEntityDao;
+          $this->tmpPath = $tmpPath;
+      }
+
+      // ...
+  }
+  ```
+
+* **Determine how your new class will be used and ensure it is constructed by the DI container.**
+
+  **If another one of your classes creates your new class...**
+
+  ... simply add the new class to the constructor of the class that's using it. If the class that's using it is stored in the container, then your new class will be constructed using DI automatically.
+
+  If the user class cannot be stored in the container, then you must use `Piwik\Container\StaticContainer::get('Piwik\Plugins\MyPlugin\MyNewClass')` to get the instance in the user class.
+
+  **If Piwik creates your new class...**
+
+  Some classes are constructed by Piwik core (for example, API classes, Report classes, etc.). Some of these classes will automatically be constructed using the DI container. For these classes, you don't have any extra work to do.
+
+  If the class is not constructed with DI, then you must use `StaticContainer` as described above.
+
+  The list of classes that are automatically constructed using DI include:
+
+  * API classes.
+  * Implementations of the [Auth interface](http://developer.piwik.org/api-reference/Piwik/Auth).
+  * `Widgets` classes.
+  * Plugin `Settings` classes.
+  * `Type` classes.
+  * Plugin `Menu` classes.
+  * Plugin `Tasks` classes.
+
+  In 3.0, all classes will be constructed with dependency injection.
+
+### Configuring the container
+
```

Comment:
This guide is mostly for core + PRO developers. 3rd party plugin developers do not actually have to use DI, and if they never write tests they're obviously never going to use DI. I can move the whole guide to piwik-in-depth or to some section for core/pro devs and revert to napoli's original guide for plugin devs.

> Eg they possibly don't need to create classes that can be changed by other plugins etc.

I know this is about plugin development, but I want to make it clear in case it's not: for core development (and probably pro development, though I don't know how much affect we have there), all classes that contain important application logic should be stored in the container. Regardless of whether there is a reason for them to be changed by plugins or not.

It is not each class' responsibility to obtain their dependencies (ie, by creating instances manually, accessing singletons or using a service locator), it is the container's responsibility to wire them. Any class that doesn't have important application logic could be replaced or stored in a 3rd party library or component.